plugins/net_plugin/net_plugin.cpp

```diff
@@ -586,10 +598,53 @@ namespace eosio {
       block_status_monitor& operator=( block_status_monitor&& ) = delete;
    };
 
+   struct flex_socket {
+      using socket_ptr     = std::shared_ptr<tcp::socket>;
+      using ssl_socket_ptr = std::shared_ptr<ssl_stream>;
+      using socket_variant = std::variant<socket_ptr, ssl_socket_ptr>;
```

Comment:
Since we already have a bool flag to identify which is which, I think the variant just adds complexity and also adds an extra level of indirection and just have a tcp_socket_ptr and a ssl_socket_ptr (and make them protected so they are not exposed). And then inside this class use ssl_socket_ptr in place of ssl_socket. (You could also skip the use_ssl flag and use if ssl_socket_ptr bool operator(), but I'm fine either way).

Or even better, I think it would be good to make flex_socket a base class with pure virtual methods and then pull in all the logic in connection that uses this into the class itself by passing lambdas into the socket.  I think @huangminghuang could help you.

Reply:
We will clean this up in a subsequent change

Reply:
I'm not sure how do you think it can be improved. variant decision looks to me more elegant than any virtual stuff. virtual stuff adds some little performance overhead and more complexity. Also `ssl_socker_ptr` and `tcp_socket_ptr` are not interchangeable as calls like boost::async_write can't accept any common interface.

Reply:
there is an alternative to always use `ssl_socket_ptr` but even in that case we won't eliminate `if/else` and usage will be like that:
`         if (my_impl->ssl_enabled){
            boost::asio::async_write( *c->socket, bufs, boost::asio::bind_executor( c->strand, write_lambda ));
         }
         else {
            boost::asio::async_write( *c->socket.lowest_layer(), bufs, boost::asio::bind_executor( c->strand, write_lambda ));
         }`
this kind of hack as we are using ssl stream class without touching any ssl logic so I decided to make it more straightforward with `variant` and use proper classes in corresponding cases.